### PR TITLE
SCT-3100: Finish Narrative SampleSet viewer changes

### PR DESCRIPTION
**Pull request by eapearson**
# Improve SampleSet viewer widget and friends

As we prepare to release samples support, this is an effort to finalize the SampleSet viewer. Specifically, it integrates some of the changes from an unmarked PR, aligns the viewer with the landing page, and adds (nearly) full test coverage.

- incorporates some of the changes from an earlier PR
- removes the SetAPI support from those changes
- aligns the sample display with the SampleSet landing page
- adds improved JSON-RPC 1.1 support with full tests
- adds Mock Service Worker (msw) for fetch-based service mocking

The changes require that the `msw` npm package be added to the project.

# Jira Ticket / Issue #
- https://kbase-jira.atlassian.net/browse/SCT-3100
- [x] Added the Jira Ticket to the title of the PR (e.g. `DATAUP-69 Adds a PR template`)

# Testing Instructions

Unit tests have been added to cover the widget, associated new support widgets, the JSON-RPC library code. In addition msw (Mock Service Worker) has been added and integrated to allow mocking of services with fetch-based clients.

In addition to unit tests, you can inspect the changes by running the Narrative locally, opening a Narrative which contains a SampleSet or in which you will generate one, and then dropping a sample set data object into the narrative. This will invoke the viewer.

- [x] Tests pass locally and in GitHub Actions
- [x] Changes available by spinning up a local narrative and navigating to _X_ to see _Y_

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/lbl.gov/trussresources/home?authuser=0
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [-] Any dependent changes have been merged and published in downstream modules
- [x] (JavaScript) I have run Prettier and ESLint on changed code manually or with a git precommit hook
- [-] (Python) I have run Black and Flake8 on changed Python code manually or with a git precommit hook

# Updating Version and Release Notes (if applicable)

- [-] [Version has been bumped](https://semver.org/) for each release
- [x] [Release notes](/RELEASE_NOTES.md) have been updated for each release (and during the merge of feature branches)

### CODE REVIEW

**Comment by eapearson** [comment]
I have yet to iterate on any issues reported by the automated PR review bots. When that is complete, I'll contact someone for review (but please feel free to start getting familiar with the changes before.)

**Comment by eapearson** [comment]
I think all the Codacy/Sonarcloud complaints are trivial or wrong, so it is ready for review.

File: kbase-extension/static/kbase/js/widgets/common/errorMessage.js [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+    /**
+     * Given a string, returns a jquery object wrapping an html tag which contains the
+     * string set, safely, via text().
+     *
+     * @param {string} text A string to render inside the given tag
+     * @param {string} tag The html tag name used to wrap the provided text; defaults to 'div'
+     * @returns JQuery object
+     */
+    function $textIn(text, tag = 'div') {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/common/errorMessage.js:66]: This seems to duplicate the `RenderIn` module from your PR #2385. I haven't gone in detail yet - does this get factored out of here, into there?

Comment (briehl) [reply kbase-extension/static/kbase/js/widgets/common/errorMessage.js:66]: If it doesn't yet, that's fine, just noting here for future work. I don't want to see this one get bigger...

File: kbase-extension/static/kbase/js/widgets/function_output/samples/kbaseSampleSet.js [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+            });
+
+            // Render
+            this.render();
+
+            return this;
+        },
+
+        loggedInCallback: function () {
+            this.render();
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/function_output/samples/kbaseSampleSet.js:94]: Having `this.render()` here and in `init` will probably do the rendering twice.

IMO, the utility of the `kbaseAuthenticatedWidget` (getting auth tokens) is pretty much already included in the Runtime module. 

Though, really, as long as `this.authToken()` works in the `init` function, you can just skip the `loggedinCallback` and save a render cycle.

**Review by briehl — CHANGES_REQUESTED** [review]
Mostly, I have a couple questions, but there's some code duplication that could be tweaked.

I really like the MockServiceWorker. Looking forward to using it!

I also like the JSONRpcClient changes and generalization. I'd like to see that replace other stuff in the Narrative, and other KBase-UI components. I think it's worth making a project out of putting that in its own NPM module (modified for use in other places, not just as an AMD). But that's another project...

File: kbase-extension/static/kbase/js/jsonrpc/1.1/errors.js [hunk 3]
```diff
@@ -0,0 +1,175 @@
+/*jslint white:true,browser:true*/
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/jsonrpc/1.1/errors.js:1]: not needed -- we no longer use jslint

File: kbase-extension/static/kbase/js/widgets/common/errorMessage.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+     *        name - string - always JSONRPCError
+     *        code - number - JSONRPC compatible error code
+     *        message - string - canonical message
+     *        error - string - captures trace
+     *
+     */
+    function renderJSONArray(data) {
+        const $rows = data.map((value, index) => {
+            return $('<tr>').append(
+                $('<th>').css('color', 'rgba(150, 150, 150, 1)').text(String(index)),
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/widgets/common/errorMessage.js:17]: No inline CSS, please.

File: .eslintrc.yml [hunk 5]
```diff
@@ -40,3 +40,7 @@ rules:
     - warn
   prefer-spread:
     - warn
+  indent:
+    - error
+    - 4
+    - SwitchCase: 1
```
Comment (ialarmedalien) [inline .eslintrc.yml:46]: what is the motivation for this change?

File: kbase-extension/static/kbase/js/widgets/common/errorMessage.js (hunk 1 above)
Comment (eapearson) [reply kbase-extension/static/kbase/js/widgets/common/errorMessage.js:66]: I think there are multiple implementations of testing utilities, due to overlapping PRs and the need for such utilities.
They should shake out over time.

File: .eslintrc.yml (hunk 5 above)
Comment (eapearson) [reply .eslintrc.yml:46]: I believe it was eslint and prettier fighting over indentation of case within switch.

**Comment by briehl** [comment]
Ok, this seems reasonable for getting on CI.

And for doing PRs next week, keep them smaller, please.

**Review by briehl — APPROVED** [review]
